### PR TITLE
Added visibility bitmask as an alternative SSAO method

### PR DESCRIPTION
**Pull request by dragostis**
Early implementation. I still have to fix the documentation and consider writing a small migration guide.

Questions left to answer:

* [x] should thickness be an overridable constant?
* [x] is there a better way to implement `Eq`/`Hash` for `SSAOMethod`?
* [x] do we want to keep the linear sampler for the depth texture?
* [x] is there a better way to separate the logic than preprocessor macros?

vbao

## Migration guide

SSAO algorithm was changed from GTAO to VBAO (visibility bitmasks). A new field, `constant_object_thickness`, was added to `ScreenSpaceAmbientOcclusion`. `ScreenSpaceAmbientOcclusion` also lost its `Eq` and `Hash` implementations.

### CODE REVIEW

File: crates/bevy_pbr/src/ssao/mod.rs
Comment (JMS55) [inline crates/bevy_pbr/src/ssao/mod.rs:465]: Why did you remove this shaderdef?

**Comment by JMS55** [comment]
References for reviewers:
* https://ar5iv.labs.arxiv.org/html/2301.11376
* https://cdrinmatane.github.io/posts/ssaovb-code

**Comment by dragostis** [comment]
I added a few questions to answer before merging this and also added a linear sampler for the depth texture since I noticed it slightly improves some of the ghosting noticeable in small objects that happens due to the low resolution mip that gets selected for far-away samples.

**Comment by JMS55** [comment]
Another good reference: https://cybereality.com/screen-space-indirect-lighting-with-visibility-bitmask-improvement-to-gtao-ssao-real-time-ambient-occlusion-algorithm-glsl-shader-implementation/

@dragostis to give you an update, I'd still like to get this merged for Bevy 0.15 if you're interested. It's just going to be a very slow process, as reviewer bandwidth is very limited, and we have a large backlog of rendering PRs. I'll try to test this out in a few weeks when I'll have more free time, and help push it forward.

**Comment by dragostis** [comment]
@JMS55, I'm happy to get this merged for 0.15!

In it's current form, it only handles AO. Adding GI as well is quite straightforward algorithm-wise, but would require extra textures and a separate denoise pass to look good.

**Comment by JMS55** [comment]
Yeah lets not do GI in this PR. Feel free to open another after if you're interested though.

To get this PR merged, I want to test this on some more realistic scenes and check how the perf/quality compares. If it's always better than the existing method, then imo we should remove the existing method and maintain just this one. That's basically the only blocker. Unfortunately I don't have a desktop for the next week or two to be able to run Sponza and other scenes.

There's also the opportunity to change the denoiser/downscale depth/noise patterns that I copied from XeGTAO, and try what the newer vismask impls are using. Up to you on that, we can also leave those in.

File: crates/bevy_pbr/src/ssao/mod.rs
Comment (DGriffin91) [inline crates/bevy_pbr/src/ssao/mod.rs:395]: The other sampler is called point because it's a nearest sampler.

**Comment by JMS55** [comment]
The code behind the vismask paper was open sourced (NOTE: Parts, including the denoiser, are from Unity and are not open-source licensed, be careful which files you reference): https://github.com/cdrinmatane/SSRT3

**Comment by JMS55** [comment]
I'm still quite busy with meshlets, but I need to find some time sometime soon to try this PR out and see how it compares to GTAO.

**Comment by dragostis** [comment]
@JMS55, sorry the late reply. This month has been a bit busy, but I expect o have a bit more time to push this through in September.

A good scene to test this IMO is [Blender's Classroom](https://www.blender.org/download/demo-files/) since it has a lot of thin geometry that showcases VBAO's strengths quite well.

**Comment by JMS55** [comment]
No worries, like I said I haven't had much time lately either.

Thanks, will try.

File: crates/bevy_pbr/src/ssao/ssao.wgsl
Comment (JMS55) [inline crates/bevy_pbr/src/ssao/ssao.wgsl:227]: I'd rather not use integers for this. Better to do `#ifdef VBAO #else` imo.

File: crates/bevy_pbr/src/ssao/mod.rs
Comment (JMS55) [inline crates/bevy_pbr/src/ssao/mod.rs:194]: This type and GTAO vs VBO need docs.

File: crates/bevy_pbr/src/ssao/mod.rs
Comment (JMS55) [inline crates/bevy_pbr/src/ssao/mod.rs:525]: I do not feel like this should be a shaderdef. I feel like we should probably use a uniform buffer / push constant here. Push constant is probably easier, and SSAO is not supported on WebGPU atm anyways.

**Review by JMS55 — CHANGES_REQUESTED** [review]
Sorry for the late review, but I finally had some time to look into this,

I tried it on Sponza and it looked good! Didn't have time to compare quality or performance to GTAO, or check the code thoroughly against the paper. I'll leave that up to other reviewers. VBAO seemed better at lower-sample counts especially though based on my very quick glances.

There's also probably a conversation to be had about if we want to support both methods, or use VBAO only. @dragostis it might help to post some example screenshots of the blender classroom comparing the two methods; I didn't have time to try and get the scene setup in bevy.

Side note to other reviewers - the SSAO docs say that it doesn't work on DirectX12, but I believe the naga issues should be fixed by now? We should test and remove the note if so.

**Comment by dragostis** [comment]
I loaded a textureless classroom into bevy and took a bunch of screenshots with varying thickness levels:

GTAO
Screenshot from 2024-09-17 20-58-43

VBAO 0.5
Screenshot from 2024-09-17 20-58-29

GTAO
Screenshot from 2024-09-17 21-00-52

VBAO 0.5
Screenshot from 2024-09-17 21-00-27

VBAO 1
Screenshot from 2024-09-17 21-00-40

GTAO
Screenshot from 2024-09-17 21-03-03

VBAO 0.25
Screenshot from 2024-09-17 21-02-29

VBAO 1
Screenshot from 2024-09-17 21-02-48

**Comment by JMS55** [comment]
Very nice! I'd be happy to remove GTAO and merge VBAO personally. I don't see a need for both.

**Comment by alice-i-cecile** [comment]
@dragostis is this ready to be reviewed now? If so, please take it out of draft :)

**Review by IceSentry — COMMENTED** [review]
I think I agree with @JMS55 on that one. We can probably just get rid of GTAO and just keep VBAO, it looks better in pretty much every case

**Comment by JMS55** [comment]
@dragostis the release candidate for bevy 0.15 is in about a week. I'd like to get this in before then. Code is mostly done, I think we just need to remove GTAO, and some other smaller stuff. If you're busy, I'd be happy to finish this PR out for you. Let me know if you have time, otherwise I'll probably do this myself to ensure it makes the release in time.

**Comment by dragostis** [comment]
@JMS55, I'll try to address the feedback and un-draft the PR tomorrow morning. Thanks for the heads up!

**Comment by JMS55** [comment]
Awesome, thanks! Excited to finally get this in!

Summary of needed changes (iirc, might be missing some):
* Remove GTAO
* Use a uniform/push constant for visbuffer config, not shaderdef
* Test on DX12 (should work), remove the docs about SSAO not being supported on DX12
* Write migration guide + changelog

**Comment by dragostis** [comment]
I've crossed everything off the list apart from the changelog. Since Bevy doesn't keep a changelog file, does this mean doing a writeup in the release blog post or is it something completely different?

**Comment by JMS55** [comment]
Nvm looks like we removed the changelog from our PR template in favor of just the migration guide.

File: crates/bevy_pbr/src/ssao/mod.rs
Comment (JMS55) [inline crates/bevy_pbr/src/ssao/mod.rs:180]: Do we still need this Hash/Eq implementation? Better to remove it if not imo.

**Review by JMS55 — CHANGES_REQUESTED** [review]
Thanks for the work! A few small suggestions, almost ready to go!

File: crates/bevy_pbr/src/ssao/mod.rs
Comment (dragostis) [reply crates/bevy_pbr/src/ssao/mod.rs:180]: My understanding was that, since this is a user-facing type, Hash/Eq might be useful.

**Review by IceSentry — APPROVED** [review]

**Review by JMS55 — APPROVED** [review]

**Comment by IceSentry** [comment]
I think the conflicts are just caused by the required components PR. Should be trivial to resolve.

File: crates/bevy_pbr/src/ssao/mod.rs
Comment (dragostis) [reply crates/bevy_pbr/src/ssao/mod.rs]: I've removed the implementation and change the pipeline key to only use the quality level instead of the whole struct.

**Comment by alice-i-cecile** [comment]
> I've crossed everything off the list apart from the changelog. Since Bevy doesn't keep a changelog file, does this mean doing a writeup in the release blog post or is it something completely different?

Yep, this will get a writeup in the blog post. I'll be pinging you over in the bevy-website repo to get your advice on how to write it :)

**Pull request merged**
